Content.Shared/Research/TechnologyDisk/Systems/TechnologyDiskSystem.cs

```diff
@@ -1,92 +1,100 @@
-﻿using System.Linq;
-using Content.Server.Popups;
-using Content.Server.Research.Systems;
-using Content.Server.Research.TechnologyDisk.Components;
+using System.Linq;
 using Content.Shared.Examine;
 using Content.Shared.Interaction;
-using Content.Shared.Random;
+using Content.Shared.Popups;
 using Content.Shared.Random.Helpers;
 using Content.Shared.Research.Components;
 using Content.Shared.Research.Prototypes;
+using Content.Shared.Research.Systems;
+using Content.Shared.Research.TechnologyDisk.Components;
+using Robust.Shared.Network;
 using Robust.Shared.Prototypes;
 using Robust.Shared.Random;
 
-namespace Content.Server.Research.TechnologyDisk.Systems;
+namespace Content.Shared.Research.TechnologyDisk.Systems;
 
 public sealed class TechnologyDiskSystem : EntitySystem
 {
-    [Dependency] private readonly PopupSystem _popup = default!;
-    [Dependency] private readonly ResearchSystem _research = default!;
-    [Dependency] private readonly IPrototypeManager _prototype = default!;
+    [Dependency] private readonly INetManager _net = default!;
+    [Dependency] private readonly IPrototypeManager _protoMan = default!;
     [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly SharedResearchSystem _research = default!;
 
-    /// <inheritdoc/>
     public override void Initialize()
     {
+        base.Initialize();
+
+        SubscribeLocalEvent<TechnologyDiskComponent, MapInitEvent>(OnMapInit);
         SubscribeLocalEvent<TechnologyDiskComponent, AfterInteractEvent>(OnAfterInteract);
         SubscribeLocalEvent<TechnologyDiskComponent, ExaminedEvent>(OnExamine);
-        SubscribeLocalEvent<TechnologyDiskComponent, MapInitEvent>(OnMapInit);
     }
 
-    private void OnAfterInteract(EntityUid uid, TechnologyDiskComponent component, AfterInteractEvent args)
+    private void OnMapInit(Entity<TechnologyDiskComponent> ent, ref MapInitEvent args)
+    {
+        // This is server-only business
+        if (!_net.IsServer)
+            return;
+
+        if (ent.Comp.Recipes != null)
+            return;
+
+        var weightedRandom = _protoMan.Index(ent.Comp.TierWeightPrototype);
+        var tier = int.Parse(weightedRandom.Pick(_random));
+
+        //get a list of every distinct recipe in all the technologies.
+        var techs = new List<ProtoId<LatheRecipePrototype>>();
+        foreach (var tech in _protoMan.EnumeratePrototypes<TechnologyPrototype>())
+        {
+            if (tech.Tier != tier)
+                continue;
+
+            techs.AddRange(tech.RecipeUnlocks);
+        }
+        techs = techs.Distinct().ToList();
```

Comment:
why not have techs be a hashset and use UnionWith to add the unlocks

Reply:
That's a good call.